Source/options.h

```diff
@@ -2,19 +2,22 @@
 
 DEVILUTION_BEGIN_NAMESPACE
 
-typedef struct DiabloOptions {
+class DiabloOptions final {
+public:
 	/** @brief Play game intro video on startup. */
 	bool bInto;
-} DiabloOptions;
+};
 
-typedef struct HellfireOptions {
+class HellfireOptions final {

```

Comment:
It's rather uncommon (and perhaps a bit confusing) to use `final` in places where no inheritance chains are intended.

Inheritance is discouraged overall (for the most part) so the `final` modifier may be sending the wrong signal here.

Reply:
> It's rather uncommon (and perhaps a bit confusing) to use final in places where inheritance is not at play.

I'm a bit confused by this. `final` is there exactly to prevent this class from being inherited. If one removes `final`, inheritance would become "possibly at play".

"Sealed by default" is considered a good practice in C# to avoid inadvertently exposing classes that are not "ready" for inheritance and to limit the design, so I was applying the same principle here.

> Inheritance is discouraged overall (for the most part) so the final modifier may be sending the wrong signal here.

But final not only discourages, but prevents it, which goes inline with what you mentioned is the recommendation (inheritance is discouraged in other languages too in favor of composition). How could this be sending the wrong signal? It sends the message that "you will not inherit from this". Isn't that what we want?

Reply:
It sends a message that I have to think about inheritance.

It's like writing "Does not contain uranium" on a cereal box :)

Reply:
Don't you think it's good to know that the class will never have parent classes and that inheritance will not be possible? That's extremely valuable to me.

I'm really not following your reasoning here. It's the same to me as saying if I mark a method with `const` then "I'll have to think about purity".

Everybody knows inheritance is a thing in C++, people will not "forget about it" just because they don't see the `final` keyword anywhere.

Can you please reconsider this one? It offers so many compile-time benefits, it would be a shame to drop it IMHO.

Reply:
> Don't you think it's good to know that the class will never have parent classes and that inheritance will not be possible? That's extremely valuable to me.

Well, this is my default assumption for all classes. It's a stylistic preference, I don't feel that strongly about it.

Reply:
unless there is an explicit reason to mark some thing as final i would strongly prefer not to add it. I find that it can cause problems when testing (how will you mock a final class), and it makes presumptions about how something will be used in the future.

I'm not sure what you mean by "ready for inheritance", sure thought might not have been put in to that explicitly, but is there known issues with doing so in all circumstances? Do you remove final only when you can prove that all possible usages will work?

Reply:
https://docs.microsoft.com/en-us/archive/blogs/ericlippert/why-are-so-many-of-the-framework-classes-sealed

https://blog.ploeh.dk/2021/03/08/pendulum-swing-sealed-by-default/

https://stackoverflow.com/a/268287/1946412

Mostly all the reasons for sealing a class in C# to me are general reasons. It is considered a good practice there by extremely experienced and known people in the community, including the previous language designer Eric Lippert.

> how will you mock a final class)

You extract an interface for it and mock the interface. For these classes here, there is obviously no need to: you just create instances of them with the values you want.

> and it makes presumptions about how something will be used in the future.

It respects YAH I: you don't know if inheritance will be needed. If it is in the future, you remove the `final`. Note that adding the keyword from a class is a breaking change, while removing it is not.

> Do you remove final only when you can prove that all possible usages will work?

You remove it once you have decided it needs to participate in an hierarchy chain.

Reply:
There is a difference between C# and C++ here.

In C#, all methods are virtual.

In C++, for classes without virtual methods, `final` is verbose and unnecessary IMO, because classes without virtual methods are almost never inherited to begin with.

Reply:
>. It's a stylistic preference

@glebm it's not stylistic, it's enforced by the compiler. It's a compile-time guarantee of what one wants.

Reply:
I would say a lot of this doesn't strictly apply since we are not shipping a library/framework. We are not accountable for support in the same way since this comes with no warranty or guaranty. Also I have a feeling many modders wont know what to do with highly restrictive code. Besides with the planed refactoring i would say our API is at version -1.0.0 and should not be relied on staying the same between two bug fixes.

Reply:
`final` removed.